virtcontainers/persist/fs/fs.go

```diff
@@ -0,0 +1,260 @@
+		return "", fmt.Errorf("sandbox container id required")
+	}
+
+	return filepath.Join(runStoragePath, fs.sandboxState.SandboxContainer), nil
+}
+
+// Dump sandboxState and containerState to disk
+func (fs *FS) Dump() (retErr error) {
+	// call registered hooks to set sandboxState and containerState
+	for _, fun := range fs.setFuncs {
+		fun(fs.sandboxState, fs.containerState)
+	}
+
+	// if error happened, destroy all dirs
+	defer func() {
+		if retErr != nil {
+			// TODO: log error
+			fs.Destroy()
+		}
+	}()
+
+	sandboxDir, err := fs.sandboxDir()
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(sandboxDir, dirMode); err != nil {
+		return err
+	}
+
+	if err := fs.lock(); err != nil {
+		return err
+	}
+	defer fs.unlock()
+
+	// persist sandbox configuration data
+	sandboxFile := filepath.Join(sandboxDir, persistFile)
+	f, err := os.OpenFile(sandboxFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileMode)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(fs.sandboxState); err != nil {
+		return err
+	}
+
+	// persist container configuration data
+	for cid, cstate := range fs.containerState {
+		cdir := filepath.Join(sandboxDir, cid)
+		if err := os.MkdirAll(cdir, dirMode); err != nil {
+			return err
+		}
+
+		cfile := filepath.Join(cdir, persistFile)
+		cf, err := os.OpenFile(cfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileMode)
+		if err != nil {
+			return err
+		}
+
+		if err := json.NewEncoder(cf).Encode(cstate); err != nil {
+			return err
+		}
+		cf.Close()
+	}
+
+	return nil
+}
+
+// Restore state for sandbox with name sid
+func (fs *FS) Restore(sid string) error {
+	if sid == "" {
+		return fmt.Errorf("restore requires sandbox id")
+	}
+
+	fs.sandboxState.SandboxContainer = sid
+	sandboxDir, err := fs.sandboxDir()
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(sandboxDir, dirMode); err != nil {
+		return err
+	}
+
+	if err := fs.lock(); err != nil {
+		return err
+	}
+	defer fs.unlock()
+
+	// get sandbox configuration from persist data
+	sandboxFile := filepath.Join(sandboxDir, persistFile)
+	f, err := os.OpenFile(sandboxFile, os.O_RDONLY, fileMode)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := json.NewDecoder(f).Decode(fs.sandboxState); err != nil {
+		return err
+	}
+
+	// walk sandbox dir and find container
+	d, err := os.OpenFile(sandboxDir, os.O_RDONLY, fileMode)
+	if err != nil {
+		return err
+	}
+	defer d.Close()
+
+	files, err := d.Readdir(-1)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+
+		cid := file.Name()
+		cfile := filepath.Join(sandboxDir, cid, persistFile)
+		cf, err := os.OpenFile(cfile, os.O_RDONLY, fileMode)
+		if err != nil {
+			return err
+		}
+
+		var cstate persistapi.ContainerState
+		if err := json.NewDecoder(cf).Decode(&cstate); err != nil {
+			return err
+		}
+		cf.Close()
+
+		fs.containerState[cid] = cstate
+	}
+	return nil
```

Comment:
`CurPersistVersion` is hardcoded, so it will always be valid. Restoring from disk can get a `99` version number, I can add some validation during restoring.

Reply:
Not sure currently. Let's keep it as it is for now :-)